src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java

```diff
@@ -1,64 +1,216 @@
-
-        final Imp modifiedFirstImp = firstImp.toBuilder().ext(mapper.mapper().valueToTree(ttxImpExt)).build();
-
-        if (modifiedImps.size() == 1) {
-            requestBuilder.imp(Collections.singletonList(modifiedFirstImp));
-        } else {
-            final List<Imp> subList = modifiedImps.subList(1, modifiedImps.size());
-            final List<Imp> finalizedImps = new ArrayList<>(subList.size() + 1);
-            finalizedImps.add(modifiedFirstImp);
-            finalizedImps.addAll(subList);
-            requestBuilder.imp(finalizedImps);
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        if (request.getImp().get(0).getBanner() == null
+                && request.getImp().get(0).getVideo() == null) {
+            errors.add(BidderError.badInput("At least one of [banner, video] "
+                    + "formats must be defined in Imp. None found"));
+            return Result.withErrors(errors);
         }
-        requestBuilder.site(modifySite(bidRequest.getSite(), firstImpExt.getSiteId()));
+        final HttpRequest<BidRequest> httpRequest = createRequest(request, errors);
+
+        return Result.of(Collections.singletonList(httpRequest), errors);
     }
 
-    private static Site modifySite(Site site, String siteId) {
-        final Site.SiteBuilder siteBuilder = site == null ? Site.builder() : site.toBuilder();
-        return siteBuilder
-                .id(siteId)
+    private HttpRequest<BidRequest> createRequest(BidRequest request, List<BidderError> errors) {
+        final BidRequest.BidRequestBuilder requestToUpdate = request.toBuilder();
+        final Imp firstImp = request.getImp().get(0);
+        try {
+            final ExtImpTtx extImpTtx = parseImpExt(firstImp);
+            requestToUpdate.site(updateSite(request.getSite(), extImpTtx.getSiteId()));
+            requestToUpdate.imp(updateImpList(request.getImp(), firstImp,
+                    extImpTtx.getProductId(), extImpTtx.getZoneId(), errors));
+        } catch (PreBidException e) {
+            errors.add(BidderError.badInput(e.getMessage()));
+        }
+
+        final BidRequest modifiedRequest = requestToUpdate.build();
+
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(HttpUtil.headers())
+                .payload(modifiedRequest)
+                .body(mapper.encode(modifiedRequest))
                 .build();
     }
 
+    private ExtImpTtx parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), TTX_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private List<Imp> updateImpList(List<Imp> requestImps, Imp firstImp, String productId,
```

Comment:
i'd suggest to divide this method in `updateFirstImp(..)` and `replaceFirstImp(imps, firstImp)`.